### PR TITLE
WINC-714: Introduce no-op WICD

### PR DESCRIPTION
**Pull request by alinaryan**
This PR handles the 'framework' creation of Windows Instance Config Daemon(WICD) by creating a no-op WICD binary and adding a new Makefile target that builds windows-instance-config-daemon.exe using GOOS=windows.

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -130,6 +130,14 @@ rules:
   - signers
```
Comment (sebsoto) [inline config/rbac/role.yaml:130]: this change is unrelated, and should be dropped from this PR

File: build/build.sh [hunk 2]
```diff
@@ -31,6 +31,7 @@ if [[ -z "$OUTPUT_DIR" ]] || [[ -z "$WMCO_SEMVER" ]] ; then
 fi
 
 PACKAGE="github.com/openshift/windows-machine-config-operator"
```
Comment (sebsoto) [inline build/build.sh:33]: this is no longer being used and can be removed

File: build/build.sh [hunk 3]
```diff
@@ -31,6 +31,7 @@ if [[ -z "$OUTPUT_DIR" ]] || [[ -z "$WMCO_SEMVER" ]] ; then
...
+MAIN_PACKAGE="${PACKAGE}/cmd/operator"
```
Comment (sebsoto) [inline build/build.sh:34]: MAIN_PACKAGE is a bit of a misnomer, this can be renamed to something like WMCO_CMD_DIR

File: Makefile [hunk 4]
```diff
@@ -92,6 +92,9 @@ build: fmt vet
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run cmd/operator/main.go
 
+daemon: fmt vet
+	go run cmd/daemon/main.go
+
```
Comment (sebsoto) [inline Makefile:97]: WICD is meant to be a windows exe, instead of having the endpoint use `go run`, please make this compile an executable called `windows-instance-config-daemon.exe`, you will need to cross compile for Windows by setting `GOOS`. the exe should be saved to the same directory that the WMCO binary is outputted to.

Comment (mansikulkarni96) [reply Makefile:97]: You can use how we [build](https://github.com/openshift/windows-machine-config-bootstrapper/blob/38466b767564bddeeb68a6a9fab708dd73a937ce/Makefile#L8) the wmcb.exe as a reference.

Comment (alinaryan) [reply Makefile]: Thanks! Fixed

File: cmd/daemon/main.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package main
```
Comment (selansen) [inline cmd/daemon/main.go:1]: should we add some file header info like below? I see it in many files but this is not followed in all files. Since this is new file, should we consider it?
``
/*
Copyright 2021.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
.....
*/

**Comment by selansen** [comment]
@aravindhp Could you pls quickly review and let us know your comments? As this is going to be our base PR to create a new WICD module, your input would greatly help

File: Makefile [hunk 6]
```diff
@@ -89,6 +90,10 @@ endif
 build: fmt vet
```
Comment (aravindhp) [inline Makefile:90]: Shouldn't doing `make build` also result in building WICD? Otherwise you will need to add another explicit entry to build this in the release repo. Adding `build-wicd` will achieve that here which will also give you a CI signal as to whether it compiles or not.

File: cmd/daemon/main.go [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+import (
+	"fmt"
+)
+
+func main() {
+	fmt.Printf("WICD placeholder...")
```
Comment (aravindhp) [inline cmd/daemon/main.go:8]: While this is fine to do for a start, we need to get the logging story right for WICD the 2nd time around. So please think through what logging mechanism will be used. 

@sebsoto if you have some particular logger in mind, I suggest to start using that from the beginning.

File: Makefile [hunk 8]
```diff
@@ -89,6 +90,10 @@ endif
...
 	build/build.sh ${OUTPUT_DIR} ${WMCO_VERSION} ${GO_MOD_FLAGS}
 
+.PHONY: build-wicd
+build-wicd:
```
Comment (aravindhp) [inline Makefile:94]: `build-wicd` or `build-daemon`? 😅

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @alinaryan. Please address my comments.

The commit `[wicd] add WICD main.go placeholder` should just be `[wicd] add main.go placeholder`. This leads me to if this should be referred to as `daemon` or `wicd`? We don't want a repeat of WMCB naming. 

The same needs to be called for the PR header itself. `[wmco] no-op WICD` is a bit strange as WICD and WMCO are independent. They just happen to live in the same repository. This PR should be something like `Introduce no-op WICD`

File: Makefile (hunk 6 above)
Comment (alinaryan) [reply Makefile:90]: Good point. I will build the daemon under `build` instead of creating a new target.

File: cmd/daemon/main.go (hunk 5 above)
Comment (alinaryan) [reply cmd/daemon/main.go:17]: Added

File: Makefile (hunk 6 above)
Comment (alinaryan) [reply Makefile:90]: Added

Comment (aravindhp) [reply Makefile:90]: > Good point. I will build the daemon under `make build` instead of creating a new target.

It is fine to have a separate target. That has the added flexibility of being able to compile just the daemon. In fact we might want to add a separate target to build the operator too.

File: cmd/daemon/main.go (hunk 7 above)
Comment (sebsoto) [reply cmd/daemon/main.go:24]: I would suggest we use klog in order to maintain uniformity with the other on-node programs

File: Makefile (hunk 6 above)
Comment (saifshaikh48) [reply Makefile:90]: +1 to separate make target

File: cmd/daemon/main.go (hunk 7 above)
Comment (alinaryan) [reply cmd/daemon/main.go:24]: @sebsoto should I start using that in this PR?

File: Makefile (hunk 6 above)
Comment (sebsoto) [reply Makefile:90]: I dont see a reason to have an entry in the release repo for this

File: cmd/daemon/main.go (hunk 7 above)
Comment (sebsoto) [reply cmd/daemon/main.go:24]: Personally I don't think that's necessary as the ticket only requires a no-op WICD and nothing else

File: Makefile (hunk 8 above)
Comment (sebsoto) [reply Makefile]: daemon is preferred for me

Comment (selansen) [reply Makefile]: I prefer daemon too :)

File: Makefile (hunk 6 above)
Comment (selansen) [reply Makefile:90]: why do we need to have an explicit entry in the release repo? Lets keep release repo change as minimal as possible.

File: cmd/daemon/main.go (hunk 7 above)
Comment (selansen) [reply cmd/daemon/main.go:24]: because this is simple PR, I suggest adding some basic logs like "process running or something" ( in that way we know log files getting created and stored in the right place). A simple one to knock it along with this PR.

Comment (alinaryan) [reply cmd/daemon/main.go]: Added

File: Makefile [hunk 9]
```diff
@@ -77,6 +77,7 @@ vet: ## Run go vet against code.
 ##@ Build
 
 OUTPUT_DIR="build/_output"
+DAEMON_DIR="cmd/daemon/main.go"
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile]: Thanks, I got rid of the var.

File: Makefile (hunk 6 above)
Comment (alinaryan) [reply Makefile:89]: I don't think it is needed either
Screenshot from 2022-03-08 10-32-24
. When running `make build-daemon` the binary is saved to the _output/bin directory.

File: cmd/daemon/main.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+	"k8s.io/klog/v2"
+)
+
+func main() {
+	klog.InitFlags(nil)
+	// By default klog writes to stderr. Setting logtostderr to false makes klog
+	// write to a log file.
+	flag.Set("logtostderr", "false")
+	flag.Set("log_file", "daemonFile.log")
```
Comment (sebsoto) [inline cmd/daemon/main.go:30]: Lets not overwrite these flags at the moment, the defaults are fine for this purpose.
WMCO will be setting the flags when it starts WICD

**Review by sebsoto — CHANGES_REQUESTED** [review]

**Comment by alinaryan** [comment]
/test vsphere-e2e-operator
Networking failure

**Comment by alinaryan** [comment]
/test vsphere-e2e-operator
Destroy cluster failure
